django/urls/constraints.py

```diff
@@ -0,0 +1,135 @@
+from __future__ import unicode_literals
+
+import re
+
+from django.core.exceptions import ImproperlyConfigured
+from django.urls.exceptions import NoReverseMatch, Resolver404
+from django.utils import six
+from django.utils.encoding import force_text
+from django.utils.functional import cached_property
+from django.utils.regex_helper import normalize
+from django.utils.translation import get_language
+
+
+class Constraint(object):
+    def describe(self):
+        """
+        Return a concise description to be used as an error message.
+        """
+        return ""
+
+    def match(self, path, request=None):
+        """
+        If this constraint matches, return (new_path, args, kwargs).
+        Otherwise, raise a Resolver404.
+        """
+        raise NotImplementedError("Constraint subclasses should implement match()")
+
+    def construct(self, url_object, *args, **kwargs):
+        """
+        (Partially) reconstruct url_object based on args and kwargs.
+        Return (url_object, leftover_args, leftover_kwargs).
+        Raise NoReverseMatch if the url cannot be reconstructed with these
+        arguments.
+        """
+        raise NotImplementedError("Constraint subclasses should implement construct()")
+
+
+class RegexPattern(Constraint):
+    def __init__(self, regex):
+        self._regex = force_text(regex)
+
+    def describe(self):
+        pattern = self.regex.pattern
+        return str(pattern[1:]) if pattern.startswith('^') else str(pattern)
```

Comment:
Historically the different regexes for includes and views are concatenated to form a single regex in all error messages and the like, and only the first start anchor is kept (see `BaseResolver.describe()`). At the moment it's a bit of an ad hoc implementation, and it doesn't account for the increased complexity of Constraints yet.